File: Matlab_code/prob_CCA/run_cca.m
% Example script to run the DASF algorithm to solve the CCA Problem.

% Author: Cem Musluoglu, KU Leuven, Department of Electrical Engineering
% (ESAT), STADIUS Center for Dynamical Systems, Signal Processing and Data
% Analytics
% Correspondence: cemates.musluoglu@esat.kuleuven.be

clear all
close all

addpath('../dasf_toolbox/');

% Number of Monte-Carlo runs.
mc_runs=5;

% Number of nodes.
nbnodes=10;
% Number of channels per node.
nbsensors_vec=5*ones(nbnodes,1);
% Number of channels in total.
nbsensors=sum(nbsensors_vec);

% Number of samples of the signals.
nbsamples=10000;

% Number of filters of X.
Q=2;

% Number of variables.
nbvariables=2;

norm_error=cell(mc_runs,1);

for n_runs=1:mc_runs

    % Create the data.
    [Y,V]=create_data(nbsensors_vec,nbnodes,nbsamples);
    
    % Structure related to the data of the problem.
    Y_cell{1}=Y;
    V_cell{1}=V;
    
    data_X=struct;
    data_X.Y_cell=Y_cell;
    data_X.B_cell={};
    data_X.Gamma_cell={};
    data_X.Glob_Const_cell={};
    
    data_W=struct;
    data_W.Y_cell=V_cell;
    data_W.B_cell={};
    data_W.Gamma_cell={};
    data_W.Glob_Const_cell={};
    
    data=cell(nbvariables,1);
    data{1}=data_X;
    data{2}=data_W;

    % Structure related to parameters of the problem.
    prob_params=struct;
    prob_params.nbsensors=nbsensors;
    prob_params.Q=Q;
    prob_params.nbnodes=nbnodes;
    prob_params.nbsensors_vec=nbsensors_vec;
    prob_params.nbsamples=nbsamples;
    prob_params.nbvariables=nbvariables;
    
    % Random updating order.
    prob_params.update_path=randperm(nbnodes);
    
    % Estimate filter using the centralized algorithm.
    [X_star,f_star]=cca_solver(prob_params,data);
    prob_params.X_star=X_star;
    % Compute the distance to X_star if "true".
    prob_params.compare_opt=true;
    % Show a dynamic plot if "true".
    prob_params.plot_dynamic=false;

    % Structure related to stopping conditions. We fix the number of 
    % iterations the DASF algorithm will perform to 1000.
    conv=struct;
    conv.nbiter=1000;
    
    % Create adjacency matrix (hollow matrix) of a random graph.
    adj=randi(2,nbnodes,nbnodes)-1;
    graph_adj=triu(adj,1)+tril(adj',-1);
    non_connected=find(all(graph_adj==0));
    % If there are non-connected nodes, connect them to every other node.
    if(non_connected)
        graph_adj(:,non_connected)=1;
        graph_adj(non_connected,:)=1;
    end
    prob_params.graph_adj=graph_adj;
    
    % Solve the CCA problem in a distributed way using the DASF framework.
    [X_est,norm_diff,norm_err,f_seq]=dasf_multivar(prob_params,data,...
                    @cca_solver,conv,@cca_select_sol,@cca_eval);
    
    norm_error{n_runs}=norm_err;
    
end

%%

% Plot the normalized error.

x_int=[1:conv.nbiter];
q_5=quantile(cell2mat(norm_error),0.5);
q_25=quantile(cell2mat(norm_error),0.25);
q_75=quantile(cell2mat(norm_error),0.75);
loglog(q_5,'b','LineWidth',2);

hold on
fill([x_int,fliplr(x_int)],[q_5,fliplr(q_75)],'b','FaceAlpha','0.2','LineStyle','none')
fill([x_int,fliplr(x_int)],[q_5,fliplr(q_25)],'b','FaceAlpha','0.2','LineStyle','none')
xlim([1,inf])
ylim([1e-8,inf])

ax=gca;
ax.FontSize=14;
xlabel('Iterations','Interpreter','latex','Fontsize',20)
ylabel('$\epsilon$','Interpreter','latex','Fontsize',20)
grid on

%%

function [Y,V]=create_data(nbsensors_vec,nbnodes,nbsamples)
    
    noisepower=0.1;
    signalvar=0.5;
    nbsources=10;
    offset=0.5;
    lags=3;

    rng('shuffle');
    D=randn(nbsources,nbsamples+lags);
    D=sqrt(signalvar)./(sqrt(var(D,0,2))).*(D-mean(D,2)*ones(1,nbsamples+lags));

    for k=1:nbnodes
        A{k}=rand(nbsensors_vec(k),nbsources)-offset;
        noise{k}=randn(nbsensors_vec(k),nbsamples+lags); 
        noise{k}=sqrt(noisepower)./sqrt(var(noise{k},0,2)).*(noise{k}...
            -mean(noise{k},2)*ones(1,nbsamples+lags)); 
    end

    column_blk=0;

    signal_cell=cell(nbnodes,1);
    Y_cell=cell(nbnodes,1);
    V_cell=cell(nbnodes,1);
    
    for k=1:nbnodes
        signal_cell{k}=A{k}*D+noise{k};
        Y_cell{k}=signal_cell{k}(:,1:nbsamples);
        V_cell{k}=signal_cell{k}(:,lags+1:end);

        Y(column_blk+1:column_blk+nbsensors_vec(k),1:nbsamples)=Y_cell{k};
        V(column_blk+1:column_blk+nbsensors_vec(k),1:nbsamples)=V_cell{k};
        column_blk=column_blk+nbsensors_vec(k);
    end

end

File: archive_v1/Python/CCA/run_cca.py
import numpy as np
import sys
import matplotlib as mpl
# Choose plot backend.
# mpl.use('macosx')
# mpl.use('Qt5Agg')
# mpl.use('TkAgg')
mpl.use('Agg')
import matplotlib.pyplot as plt
from time import sleep

sys.path.append('../dasf_toolbox/')
import cca_functions as cca
from dasf_toolbox import dasf_multivar

# Number of Monte-Carlo runs.
mc_runs = 5

# Number of nodes.
nbnodes = 10
# Number of channels per node.
nbsensors_vec = 5 * np.ones(nbnodes)
nbsensors_vec = nbsensors_vec.astype(int)
# Number of channels in total.
nbsensors = np.sum(nbsensors_vec)

# Number of samples of the signals.
nbsamples = 10000

# Number of filters of X.
Q = 2

# Number of variables.
nbvariables = 2

norm_error = []

rng = np.random.default_rng()

for k in range(mc_runs):
    # Create the data.
    Y, V = cca.create_data(nbsensors, nbsamples)

    # Dictionary related to the data of the problem.
    Y_list = [Y]
    V_list = [V]
    Gamma_list = []
    B_list = []
    Glob_Const_list = []

    data_X = {'Y_list': Y_list, 'B_list': B_list,
            'Gamma_list': Gamma_list, 'Glob_Const_list': Glob_Const_list}
    data_W = {'Y_list': V_list, 'B_list': B_list,
              'Gamma_list': Gamma_list, 'Glob_Const_list': Glob_Const_list}

    data = [data_X, data_W]

    # Dictionary related to parameters of the problem.
    prob_params = {'nbnodes': nbnodes, 'nbsensors_vec': nbsensors_vec,
                   'nbsensors': nbsensors, 'Q': Q, 'nbsamples': nbsamples, 'nbvariables': nbvariables}

    # Create adjacency matrix (hollow matrix) of a random graph.
    graph_adj = rng.integers(0, 1, size=(nbnodes, nbnodes), endpoint=True)
    graph_adj = np.triu(graph_adj, 1) + np.tril(graph_adj.T, -1)
    prob_params['graph_adj'] = graph_adj

    # Random updating order.
    update_path = rng.permutation(range(nbnodes))
    prob_params['update_path'] = update_path

    # Estimate filter using the centralized algorithm.
    X_star = cca.cca_solver(prob_params, data, X0=None, solver_params=None)
    f_star = cca.cca_eval(X_star, data)

    prob_params['X_star'] = X_star
    # Compute the distance to X_star if "True".
    prob_params['compare_opt'] = True
    # Show a dynamic plot if "True".
    prob_params['plot_dynamic'] = False

    # Dictionary related to stopping conditions. We fix the number of iterations the DASF algorithm will perform to 200.
    nbiter = 1000
    conv = {'nbiter': nbiter}

    # Solve the CCA in a distributed way using the DASF framework.
    X_est, norm_diff, norm_err, f_seq = dasf_multivar(prob_params, data, cca.cca_solver,
                                             conv, solver_params=None, prob_select_sol=cca.cca_select_sol, prob_eval=cca.cca_eval)

    norm_error.append(norm_err)

    sys.stdout.write('\r')
    j = (k + 1) / mc_runs
    sys.stdout.write("[%-20s] %d%%" % ('='*int(20*j), 100*j))
    sys.stdout.flush()
    sleep(0.25)

sys.stdout.write('\n')


# Plot the normalized error.
q5 = np.quantile(norm_error, 0.5, axis=0)
q25 = np.quantile(norm_error, 0.25, axis=0)
q75 = np.quantile(norm_error, 0.75, axis=0)
iterations = np.arange(1, nbiter + 1)

fig = plt.figure()
ax = fig.add_subplot(1, 1, 1)
ax.loglog(iterations, q5, color='b')
ax.fill_between(iterations, q25, q75)
ax.set_xlabel('Iterations')
ax.set_ylabel('Normalized error')
ax.grid(True, which='both')
plt.show()
plt.savefig("cca_convergence.pdf")
